heartbeat/scheduler/throttler/throttler.go

```diff
@@ -0,0 +1,118 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package throttler
+
+import (
+	"math"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common/atomic"
+)
+
+const (
+	stateUnstarted = iota
+	stateStarted
+	stateStopped
+)
+
+// Throttler is useful for managing access to some resource that can handle a certain amount of concurrency only.
+// You could also do this with a Pool, but this uses a constant amount of memory, and doesn't need to have token
+// objects passed around which is cleaner.
+type Throttler struct {
+	limit   uint
+	buf     chan struct{}
+	done    chan struct{}
+	state   atomic.Int
+	startWg sync.WaitGroup
+}
+
+// NewThrottler returns a new *Throttler that is not yet started. You must invoke Start for it to do anything.
+func NewThrottler(limit uint) *Throttler {
+	if limit < 1 { // assume unlimited
+		limit = math.MaxUint32
+	}
+
+	t := &Throttler{
+		limit:   limit,
+		buf:     make(chan struct{}, limit),
+		done:    make(chan struct{}),
+		state:   atomic.MakeInt(stateUnstarted),
+		startWg: sync.WaitGroup{},
+	}
+
+	t.startWg.Add(1)
+
+	return t
+}
```

Comment:
This looks like a simple counting semaphore would do the job. Any plans to add more functionality to the 'throttler'?

In case all we need is a semaphore, we can use this: https://godoc.org/golang.org/x/sync/semaphore. Block on start can be created by adjusting the count via Acquire and Release on init.